### ISSUE TITLE
Application w/esm using dependency w/esm must be on same version

### ISSUE THREAD

**Issue by CacheControl** [standard-things/esm#368]
Hi @jdalton 

We've started using `esm` throughout many applications and internal packages at my organization, and a recurring pain point is the need to have the application and all dependencies use the same version of `esm`.  If the version of `esm` differs between app/dependency, very often the result is a syntax error:

```bash
import './newrelic';
^^^^^^

SyntaxError: esm@3.0.16 is not enabled for <path>
```

To cite the most recent example, the above occurred when a dependency was on `esm@^3.0.19`, and the application was on `esm@^3.0.16`; not a huge gap, with the versions being separated only by a few weeks.

This issue has been impacting us every few versions of esm, to the point where we're questioning whether using the library within dependencies is worthwhile; while upgrading `esm` within an application is a trivial exercise, needing to also upgrade all `esm` based dependencies is considerably more overhead. 

Can you provide some insight into whether this is a fundamental issue that can be expected reoccur with most new versions, or should we expect this sort of pain to subside as the library matures?

**Comment by jdalton** [comment]
At the moment the `esm` loader uses the dependency version as an opt-in gate. So by saying you depend on `esm@^3.0.19` you exclude the lower `esm@^3.0.16` range since `3.0.16+` is less than `3.0.19+`. What I can do though is look for the `^` and assume any `3.x` is allowed.

**Comment by jdalton** [comment]
In other words it's not a fundamental issue just a the version gating mechanism we're using at the moment. It can totally, and should, be tweaked to support your usage.

**Comment by jdalton** [comment]
Yep, that's what it is:
```js
SemVer.maxSatisfying(["3.0.16"], "^3.0.19")
// => null
SemVer.maxSatisfying(["3.0.20"], "^3.0.19")
// => '3.0.20'
```

**Comment by CacheControl** [comment]
aha, great!  Yes that would definitely solve our pain!

**Comment by jdalton** [comment]
@CacheControl can you run a test build of this to see if it works for you?

**Comment by CacheControl** [comment]
@jdalton thanks!  yep, I'll give it a go

**Comment by Tundon** [comment]
So this is just my thought, aren’t we only supposed to use esm as a dev dependency? Only the application use esm to load all the files, the lib will only provide ES module files.

Yun

**Comment by jdalton** [comment]
>  aren’t we only supposed to use esm as a dev dependency

Nope. `esm` is intended for production use _(a direct dependency)_.

**Comment by jdalton** [comment]
That's the _"production ready"_ bit of the tag line.
> A fast, production ready, zero-dependency ES module loader for Node 6+!

**Comment by CacheControl** [comment]
@jdalton Can you provide some guidance on how to test this?

I cloned down master with the fix, and tried to run `npm run build`, and received this error:
```bash
localhost:~/esm|master⇒  npm run build

> esm@3.0.19 build /Users/chamm/esm
> node script/build.js

Invalid configuration object. Webpack has been initialised using a configuration object that does not match the API schema.
 - configuration has an unknown property 'optimization'. These properties are valid:
   object { amd?, bail?, cache?, context?, dependencies?, devServer?, devtool?, entry, externals?, loader?, module?, name?, node?, output?, parallelism?, performance?, plugins?, profile?, recordsInputPath?, recordsOutputPath?, recordsPath?, resolve?, resolveLoader?, stats?, target?, watch?, watchOptions? }
   For typos: please correct them.
   For loader options: webpack 2 no longer allows custom properties in configuration.
     Loaders should be updated to allow passing options via loader options in module.rules.
     Until loaders are updated one can use the LoaderOptionsPlugin to pass these options to the loader:
     plugins: [
       new webpack.LoaderOptionsPlugin({
         // test: /\.xxx$/, // may apply this only for some modules
         options: {
           optimization: ...
         }
       })
     ]
 - configuration.module.rules[0] has an unknown property 'type'. These properties are valid:
   object { enforce?, exclude?, include?, issuer?, loader?, loaders?, oneOf?, options?, parser?, query?, resource?, resourceQuery?, compiler?, rules?, test?, use? }
   -> A rule
{ Error: Command failed: webpack --hide-modules
    at makeError (/Users/chamm/esm/node_modules/execa/index.js:172:9)
    at Promise.all.then.arr (/Users/chamm/esm/node_modules/execa/index.js:277:16)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
  code: 1,
  stdout: null,
  stderr: null,
  failed: true,
  signal: null,
  cmd: 'webpack --hide-modules',
  timedOut: false,
  killed: false }
```

For the heck of it I went ahead and `npm link`d `esm` in my application and received the following error (assume because the build failed):
```bash
/Users/chamm/esm/esm.js:189
defineProperty(makeRequireFunction, shared.symbol.package, {
                                                  ^

TypeError: Cannot read property 'package' of undefined
    at Object.<anonymous> (/Users/chamm/esm/esm.js:189:51)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at requires.forEach.mod (/Users/chamm/sim-sdk/test/util/canaries/sim-ingester-release/node_modules/mocha/bin/_mocha:467:3)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/chamm/sim-sdk/test/util/canaries/sim-ingester-release/node_modules/mocha/bin/_mocha:466:10)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Function.Module.runMain (module.js:693:10)
    at startup (bootstrap_node.js:188:16)
    at bootstrap_node.js:609:3
npm ERR! Test failed.  See above for more details.
```

**Comment by jdalton** [comment]
No worries. I've prebuilt it as [esm-3.0.20.zip](https://github.com/standard-things/esm/files/1924932/esm-3.0.20.zip).

**Comment by jdalton** [comment]
Ah, missed a critical bit to the patch. Here is a new build 
[esm-3.0.20.zip](https://github.com/standard-things/esm/files/1925014/esm-3.0.20.zip).

**Comment by jdalton** [comment]
Sorry, tests are still failing. One second.

**Comment by CacheControl** [comment]
Shall I use the latest build you linked to test, or hold off for a new one?

**Comment by jdalton** [comment]
The third time's the charm: [esm-3.0.20.zip](https://github.com/standard-things/esm/files/1925157/esm-3.0.20.zip)

**Comment by jdalton** [comment]
Related `semver` bug: https://github.com/npm/node-semver/issues/236

**Comment by CacheControl** [comment]
@jdalton I've done some testing and discussion, and I think I need to apologize for filing a non-issue. 

We had several packages which were not playing nice with applications on various versions of esm (3.0.6, 3.0.8, 3.0.16, 3.0.19, among others) - a variety of different syntax errors were being thrown depending on the esm version combination and dependencies.  These errors were encountered by multiple different team members, and so there was a general lack of trust that had developed recently around mixing esm versions.

The specific issue I filed above was due to troubleshooting to fix a syntax error in an `esm@3.0.16` application with dependencies on several different esm versions.  Trying different versions of the application's `esm` version in an attempt to identify one that worked for all dependencies, I was using the `--no-save` option of `npm install`.  **That eventually resulted in the package.json having a higher version number of `esm` than was actually installed**, resulting in the error message posted above.  My guess is the SemVer library esm uses references the content of package.json.

The syntax errors we encountered were fixed by your new build, but I'm fairly certain that's due to the upgrade itself, as installing `3.0.19` works as well.

Again, apologies!
